docs/HOWTO-scale-spoke-plan.md

```diff
@@ -0,0 +1,172 @@
+* dynamicAssignment -- for dyanmic assignment-enabled campaigns, allowing a texter to 'take' a queue of contacts for assignment and begin sending.
+
+#### DB write queue
+
+While the Redis cache layer will suffice for the real-time read/write
+interface during high traffic, we still want to ensure that the DB is
+eventually synchronized.  This should be done with a message queue
+that synchronizes the database (eventually).  Making this separate
+from Redis will allow for fault tolerance, but also adds an additional
+technology to be integrated and maintained.  This work may align with
+other interest in making the existing Spoke job queue (to handle,
+e.g. texter assignment, contact uploads) -- but might not?
+
+Some technologies that could support this:
+* Amazon SQS
+* Heroku AMQP: https://elements.heroku.com/addons/cloudamqp
+* Should we have a 'backup option' that is 'pure redis' so folks can use that optionally?
+
+
+
+#### Redis Data-structures
+
+Besides satisfying the above data needs/write-workflows, keys should expire
+naturally when texters disengage (or go to bed :-)
+
+Here is the (proposed) structure of data in Redis to support the above data needs (c_id is "contact id" and `message_service_id` is probably global, but maybe related to the organization in a multi-tenant configuration):
+
+* HASH: `texterinfo-<texter_id>` (access
+
+  Keys: {auth0_id, `<org_id>`=`<role>`, is_superadmin}
+
+* HASH: `replies-<texter_id>-<campaign_id>`
+
+  Keys are the incoming message_id, with values with serialized (?JSON) content: {contact_cell, c_id}
+
+* QUEUE: `conversation-<contact_cell>-<message_service_id>` -- the list of messages for a contact, both sent and received
+
+* HASH: `contactinfo-<contact_cell>-<message_service_id>`
+
+  Keys with values include {assigned texter_id, assignment_id, org_id, questionResponseValues, [contact info including, e.g. city/state]}
+
+* QUEUE: `newassignments-<texter_id>-<campaign_id>` -- full contact info for all new assignments (status=needsMessage)
+
+* HASH: `unsentassigned-<texter_id>-<campaign_id>`
+
+  Keys are `<contact_cell>`; values are full contact info and assignment id/status
+  (same as values in `newassignments-<texter_id>-<campaign_id>` above)
+
+* QUEUE: `dynamicassignments-<campaign_id>` -- full contact info for all contacts ready for dynamic assignment in a campaign
+
+* KEY (regular `SET` call): `campaign-<campaign_id>` -- campaign data that is loaded in TexterTodo and TexterTodoList components
+
+#### Workflows using Data-structures
+
+* Access Control
+
+  1. Load from `texterinfo-<texter-id>`
+
+* getNew (needsMessage)
+
+  1. Load (and LPOP) from `newassignments-<texter_id>-<campaign_id>`
+  2. If newassignments is empty, we check `unsentassigned-<texter_id>-<campaign_id>` for content.  If it does, we ?resend that data with an additional setting in the API getNew so the client knows that the 'original queue' is empty.
```

Comment:
I dont understand the concept of unsentassigned-texter_id-campaign_id. When is this created? When we initially create the campaign (like in updateAssignments?).  Is the idea that newassignments is first created, and once the person starts texting, we create a new thing called unsent.. which keeps track of message status? 

Does newassignments get transformed to unsentassigned?

Reply:
So the use-case that `unsentassigned-` helps us with is imagine the following situation:
1. Some things are in newassignments
2. A texter's browser sends the query to get some new assignments.  So the system sends them some items from newassignments and removes them from newassignments (so that future queries won't re-send the same ones)
3. The texter leaves, a network failure occurs, or the texter's browser crashes.
4.  Finally later the texter logs in again to resume.  Their browser does more queries to newassignments, and they finish their task

The problem here is that those middle contacts got dropped.  By storing unsentassigned, we track which ones are 'in the queue' -- i.e. sent to the texter, but the texter has not yet actually sent them a message/response.  Thus we only want to fully remove them after the texter *sends* them a response.  Until then, they're in 'purgatory'--i.e. we'll skip them until the texter thinks they're finished.